### PR TITLE
Cosmos: Implement SelectMany

### LINKED ISSUE #17312: Cosmos: Translate SelectMany

### PR DESCRIPTION
**Pull request by roji**
**NOTE: This PR is based on top of #33998, skip first two commits**

Also introduces substantial infrastructure for general joins; actual Join operator support to be introduced shortly.

Closes #17312

### CODE REVIEW

File: src/EFCore.Cosmos/Properties/CosmosStrings.resx [hunk 1]
```diff
@@ -123,6 +123,9 @@
   <data name="CanConnectNotSupported" xml:space="preserve">
     <value>The Cosmos database does not support 'CanConnect' or 'CanConnectAsync'.</value>
   </data>
+  <data name="ComplexProjectionInSubqueryNotSupported" xml:space="preserve">
+    <value>Complex projections in subqueries are currently unsupported.</value>
```
Comment (ajcvickers) [inline src/EFCore.Cosmos/Properties/CosmosStrings.resx:127]: A second sentence on what they can do about this?

File: src/EFCore.Cosmos/Query/Internal/CosmosQuerySqlGenerator.cs [hunk 2]
```diff
@@ -283,16 +283,11 @@ protected override Expression VisitSelect(SelectExpression selectExpression)
...
             // Cosmos projects a JSON object containing the value).
-            if (selectExpression.UsesSingleValueProjection)
+            // TODO: Ideally, just always use VALUE for all single-projection SELECTs - but this like requires shaper changes.
+            if (selectExpression.UsesSingleValueProjection && projection is [var singleProjection])
             {
                 _sqlBuilder.Append("VALUE ");
 
-                if (projection is not [var singleProjection])
-                {
-                    throw new UnreachableException(
```
Comment (ajcvickers) [inline src/EFCore.Cosmos/Query/Internal/CosmosQuerySqlGenerator.cs:240]: Let's discuss UnreachableExceptions in a design meeting. :-)

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -1000,16 +1000,52 @@ protected override ShapedQueryExpression TranslateSelect(ShapedQueryExpression s
...
-        => null;
+    {
+        // TODO: Note that we currently never actually seem to get SelectMany without a result selector, because nav expansion rewrites
+        // that to a more complex variant with a result selector (see https://github.com/dotnet/efcore/issues/32957#issuecomment-2170950767)
+        // blogs.SelectMany(c => c.Ints) becomes:
+        // blogs
+        //     .SelectMany(p => Property(p, "Ints").AsQueryable(), (p, c) => new TransparentIdentifier`2(Outer = p, Inner = c))
+        //     .Select(ti => ti.Inner)
+
+        // TODO: In Cosmos, we currently always add a predicate for the discriminator, so the source is never a bare array.
```
Comment (ajcvickers) [inline src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:1039]: You can always use `HasNoDiscriminator` in tests.

Meta-point: too many of our tests have:

- Discriminators by default
- No partition key
- No owned types

I think this means we're largely testing un-representative models.

File: src/EFCore.Cosmos/Query/Internal/Expressions/SelectExpression.cs [hunk 4]
```diff
@@ -517,6 +510,86 @@ public virtual void ReverseOrderings()
...
+            }
+            else
+            {
+                // TODO: #34004
+                // The subquery is projecting out a JSON object; for the projection mapping of the outer query, we need to generate
+                // property accesses over that object: Scalar/ObjectAccessExpressions over the ObjectReferenceExpression that references
+                // the JOIN source.
+                // However, the JSON object being projected out of the subquery doesn't correspond to any entity type, and there's currently
+                // no way for us to represent a reference to that - ObjectReferenceExpression requires an IEntityType. Changing that
+                // requires shaper-side changes (see comment in ObjectReferenceExpression); if we can remove that requirement, we can
```
Comment (ajcvickers) [inline src/EFCore.Cosmos/Query/Internal/Expressions/SelectExpression.cs:573]: 👍

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore.Cosmos/Properties/CosmosStrings.resx (hunk 1 above)
Comment (roji) [reply src/EFCore.Cosmos/Properties/CosmosStrings.resx:127]: Unfortunately I'm not sure what can be done :(

I think we should try to fix this for 9.0 - that's #34004, and I think it actually shouldn't be too hard. But it does involve some shaper changes...

Let's discuss this (but I'll go ahead and merge for now).

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:1039]: Good point, with HasNoDiscriminator() there's indeed no predicate. But unfortunately there's still the problem described just above, where nav expansion rewrites the simple SelectMany() to the complete form with a result selector, and an additional Select that projects the inner out. So doing the simplification here to remove the JOIN is pretty much unfeasible until we fix that...

Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:1039]: > Meta-point: too many of our tests have [...]

Yeah, you're right. For the discriminators, we've been recently discussing changing our default mapping strategy, i.e. no longer mapping different entity types to the same container by default; if go decide to do that, presumably there won't be a discriminator by default in any case (i.e. Cosmos will be more similar to relational), so at least that point maybe goes away.

**Pull request merged**